### PR TITLE
Orca: Add type hint for tf estimator.

### PR DESCRIPTION
**Pull request by lalalapotter**
## Description

This PR is to add type hint for tf estimator.

### CODE REVIEW

**Review by leonardozcm — COMMENTED** [review]
And fix issues in github actions first.

**Comment by leonardozcm** [comment]
@jenniew hi, would you mind taking a look at this for i'm not that familar with this part.

File: python/orca/src/bigdl/orca/learn/tf/estimator.py [hunk 1]
```diff
@@ -36,6 +54,18 @@
...
+from bigdl.orca.data.shard import SparkXShards
+from bigdl.orca.data.tf.data import TensorSliceDataset
+from bigdl.orca.learn.optimizers import Optimizer
+from bigdl.orca.learn.metrics import Metric
+from bigdl.orca.learn.trigger import SeveralIteration
+from bigdl.orca.tfpark.tf_dataset import (
+    DataFrameDataset,
+    TFDataDataset,
+    TFNdarrayDataset,
+)
```
[the remark this answers is not in the document]
Comment (lalalapotter) [reply python/orca/src/bigdl/orca/learn/tf/estimator.py]: fixed.

File: python/orca/src/bigdl/orca/learn/tf/estimator.py [hunk 2]
```diff
@@ -292,10 +322,21 @@ def load_orca_checkpoint(self, path, version=None):
...
+        loss: Optional[tf.Tensor]=None,
+        optimizer: Optional[Optimizer]=None,
+        metrics: Optional[Metric]=None,
+        clip_norm: Optional[float]=None,
+        clip_value: Union[float, Tuple[float, float], None]=None,
+        updates: Optional[List[tf.Variable]]=None,#not sure
+        sess: Optional[tf.Session]=None,
+        model_dir: Optional[str]=None,
+        backend: str="bigdl"
+        ) -> TensorFlowEstimator:
```
[the remark this answers is not in the document]
Comment (lalalapotter) [reply python/orca/src/bigdl/orca/learn/tf/estimator.py]: fixed.

File: python/orca/src/bigdl/orca/learn/tf/estimator.py [hunk 3]
```diff
@@ -607,10 +661,13 @@ def fit(self, data,
                                    checkpoint_trigger=checkpoint_trigger)
         return self
 
-    def predict(self, data, batch_size=4,
-                feature_cols=None,
-                auto_shard_files=False,
-                ):
+    def predict(
+        self,
+        data: Union["SparkXShards", "DataFrame", "tf.data.Dataset"],
```
Comment (sgwhat) [inline python/orca/src/bigdl/orca/learn/tf/estimator.py:666]: Need to double check if predict supports `tf.data.Dataset`, maybe just comments mentioned.

File: python/orca/src/bigdl/orca/learn/tf/estimator.py [hunk 4]
```diff
@@ -1029,7 +1103,7 @@ def get_model(self):
         return self.model.model
 
     @enable_multi_fs_save
-    def save(self, model_path, overwrite=True):
+    def save(self, model_path, overwrite=True) -> None:
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf/estimator.py:1106]: type hint here?

File: python/orca/src/bigdl/orca/learn/tf/tf_estimator.py [hunk 5]
```diff
@@ -14,22 +14,27 @@
...
-from bigdl.orca.ray import RayContext
-from bigdl.orca.learn.tf.tf_runner import TFRunner
 import ray
+
+from bigdl.orca.ray import OrcaRayContext
+from bigdl.orca.learn.tf.tf_runner import TFRunner
 from bigdl.orca.learn.dl_cluster import RayDLCluster
+from typing import TYPE_CHECKING, Any
+if TYPE_CHECKING:
+    from typing import Dict, Optional, Callable
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf/tf_estimator.py:24]: put `Any ` under `if TYPE_CHECKING` or move line24 out

File: python/orca/src/bigdl/orca/learn/tf/tf_estimator.py [hunk 6]
```diff
@@ -126,8 +131,8 @@ def __init__(self,
         self.num_workers = len(self.remote_workers)
 
     def train_and_evaluate(self,
-                           train_spec,
-                           eval_spec):
+                           train_spec: "TrainSpec",
+                           eval_spec: "EvalSpec") -> Any:
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf/tf_estimator.py:135]: According to comments this may return a tuple of what?

**Review by leonardozcm — APPROVED** [review]
LGTM

File: python/orca/src/bigdl/orca/learn/tf/tf_estimator.py (hunk 5 above)
Comment (lalalapotter) [reply python/orca/src/bigdl/orca/learn/tf/tf_estimator.py:24]: if I put `Any` under `if TYPE_CHECKING`, UT will report error: `"no module named Any."`

Comment (lalalapotter) [reply python/orca/src/bigdl/orca/learn/tf/tf_estimator.py:24]: What if I just remove line 24 and add it in line 22?

Comment (leonardozcm) [reply python/orca/src/bigdl/orca/learn/tf/tf_estimator.py:24]: > What if I just remove line 24 and add it in line 22?

That sounds good!

File: python/orca/src/bigdl/orca/learn/tf/estimator.py (hunk 3 above)
Comment (lalalapotter) [reply python/orca/src/bigdl/orca/learn/tf/estimator.py]: Deleted.

**Pull request merged**